common/txmgr/address_state.go

```diff
@@ -238,3 +345,151 @@ func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveIn
 func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) MoveConfirmedToUnconfirmed(attempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) error {
 	return nil
 }
+
+// Close releases all resources held by the address state.
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Close() {
+	clear(as.idempotencyKeyToTx)
+
+	as.unstartedTxs.Close()
+	as.unstartedTxs = nil
+	as.inprogressTx = nil
+
+	clear(as.unconfirmedTxs)
+	clear(as.confirmedMissingReceiptTxs)
+	clear(as.confirmedTxs)
+	clear(as.allTxs)
+	clear(as.fatalErroredTxs)
+
+	as.idempotencyKeyToTx = nil
+	as.unconfirmedTxs = nil
+	as.confirmedMissingReceiptTxs = nil
+	as.confirmedTxs = nil
+	as.allTxs = nil
+	as.fatalErroredTxs = nil
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Abandon() {
+	as.Lock()
+	defer as.Unlock()
+
+	for as.unstartedTxs.Len() > 0 {
+		tx := as.unstartedTxs.RemoveNextTx()
+		as.abandonTx(tx)
+	}
+
+	if as.inprogressTx != nil {
+		tx := as.inprogressTx
+		as.abandonTx(tx)
+		as.inprogressTx = nil
+	}
+	for _, tx := range as.unconfirmedTxs {
+		as.abandonTx(tx)
+	}
+
+	clear(as.unconfirmedTxs)
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) abandonTx(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	if tx == nil {
+		return
+	}
+
+	tx.State = TxFatalError
+	tx.Sequence = nil
+	tx.Error = null.NewString("abandoned", true)
+
+	as.fatalErroredTxs[tx.ID] = tx
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) applyToTxs(
+	txIDsToTx map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	fn func(*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]),
+	txIDs ...int64,
+) {
+	// if txIDs is not empty then only apply the filter to those transactions
+	if len(txIDs) > 0 {
+		for _, txID := range txIDs {
+			tx := txIDsToTx[txID]
+			if tx != nil {
+				fn(tx)
+			}
+		}
+		return
+	}
+
+	// if txIDs is empty then apply the filter to all transactions
+	for _, tx := range txIDsToTx {
+		fn(tx)
+	}
+}
+
+func (as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) findTxAttempts(
+	txIDsToTx map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	txFilter func(*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool,
+	txAttemptFilter func(*txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool,
+	txIDs ...int64,
+) []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE] {
+	as.RLock()
+	defer as.RUnlock()
```

Comment:
I think the method that calls this also acquires the lock. Since they're both read lock, this probably isn't causing an issue right now. But in case we ever call this function in a method that makes updates (uses a write lock), we could run into deadlock. I think we should remove the lock from here but clearly call out in the comments that it's unsafe and expects the lock to already be acquired by the caller.